forge-game/src/main/java/forge/game/GameSnapshot.java

```diff
@@ -0,0 +1,458 @@
+package forge.game;
+
+import java.util.*;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import forge.game.card.*;
+import forge.game.combat.Combat;
+import forge.game.mana.Mana;
+import forge.game.phase.PhaseHandler;
+import forge.game.player.Player;
+import forge.game.player.RegisteredPlayer;
+import forge.game.spellability.SpellAbility;
+import forge.game.spellability.SpellAbilityStackInstance;
+import forge.game.trigger.TriggerType;
+import forge.game.zone.PlayerZoneBattlefield;
+import forge.game.zone.ZoneType;
+
+public class GameSnapshot {
+    private final Game origGame;
+    private Game newGame = null;
+    private boolean restore = false;
+
+    private final SnapshotEntityMap gameObjectMap = new SnapshotEntityMap();
+
+    public GameSnapshot(Game origGame) {
+        this.origGame = origGame;
+    }
+
+    public Game getCopiedGame() {
+        return origGame;
+    }
+
+    public Game makeCopy() {
+        return makeCopy(null, true);
+    }
+    public Game makeCopy(List<RegisteredPlayer> replacementPlayers, boolean includeStack) {
+        List<RegisteredPlayer> newPlayers;
+        if (replacementPlayers != null) {
+            newPlayers = replacementPlayers;
+        } else {
+            // Create new RegisteredPlayers based off original Match RPs
+            newPlayers = Lists.newArrayList(origGame.getMatch().getPlayers());
+        }
+        GameRules currentRules = origGame.getRules();
+        Match newMatch = new Match(currentRules, newPlayers, origGame.getView().getTitle());
+        newGame = new Game(newPlayers, currentRules, newMatch);
+        restore = false;
+        assignGameState(origGame, newGame, includeStack);
+        //System.out.println("Storing game state with timestamp of :" + origGame.getTimestamp());
+
+        return newGame;
+    }
+
+    public void restoreGameState(Game currentGame) {
+        System.out.println("Restoring game state with timestamp of :" + newGame.getTimestamp());
+        restore = true;
```

Comment:
Hmn, I think for now there should probably be some kind of rundimentary logic that does not allow the same rollback to happen twice?
This can happen quickly during payment chains, e. g.
tapping a basic to pay for mana ability and then trying to use _Ashnod's Altar_ for the rest but in the end deciding to cancel that one
-> now you're rolling back to the wrong state and the basic gets untapped

Reply:
Do you just mean that canceling any part of a cost payment is rolling back the entire thing? That's intentional for now. I think we should be a bit more precise in the future, but I'd rather just rewind the whole thing and start over, rather than leave you half paid. Ideally, it would be a stack of costs, but when i was trying to do that it was causing problems with unfreezing the stack.

Reply:
Hmn well currently the first cancel on top of another rolls back everything and leaves the game in an illegal state
that it applies the same state a second time after probably doesn't hurt?
So while this make even unintentional "cheating" rather easier to occur, that's not a huge concern for experimental? 🤔